lib/Target/DebugInfo/EmitHGLDD.cpp

```diff
@@ -253,51 +435,360 @@ void FileEmitter::emitInstance(llvm::json::OStream &json,
+                              .Case<comb::ParityOp>([](auto) { return "^"; })
+                              .Default([](auto) { return ""; });
+  if (!unaryOpcode.empty() && op->getNumOperands() == 1) {
+    auto arg = emitExpression(op->getOperand(0));
+    if (!arg)
+      return {};
+    return {JObject({
+                {"opcode", unaryOpcode},
+                {"operands", JArray{arg.expr}},
+            }),
+            result.getType()};
+  }
+
+  StringRef binaryOpcode =
+      TypeSwitch<Operation *, StringRef>(op)
+          .Case<comb::AddOp>([](auto) { return "+"; })
+          .Case<comb::SubOp>([](auto) { return "-"; })
+          .Case<comb::MulOp>([](auto) { return "*"; })
+          .Case<comb::DivUOp, comb::DivSOp>([](auto) { return "/"; })
+          .Case<comb::ModUOp, comb::ModSOp>([](auto) { return "%"; })
+          .Case<comb::ShlOp>([](auto) { return "<<"; })
+          .Case<comb::ShrUOp>([](auto) { return ">>"; })
+          .Case<comb::ShrSOp>([](auto) { return ">>>"; })
+          .Case<comb::ICmpOp>([](auto cmpOp) -> StringRef {
+            switch (cmpOp.getPredicate()) {
+            case comb::ICmpPredicate::eq:
+              return "==";
+            case comb::ICmpPredicate::ne:
+              return "!=";
+            case comb::ICmpPredicate::ceq:
+              return "===";
+            case comb::ICmpPredicate::cne:
+              return "!==";
+            case comb::ICmpPredicate::weq:
+              return "==?";
+            case comb::ICmpPredicate::wne:
+              return "!=?";
+            case comb::ICmpPredicate::ult:
+            case comb::ICmpPredicate::slt:
+              return "<";
+            case comb::ICmpPredicate::ugt:
+            case comb::ICmpPredicate::sgt:
+              return ">";
+            case comb::ICmpPredicate::ule:
+            case comb::ICmpPredicate::sle:
+              return "<=";
+            case comb::ICmpPredicate::uge:
+            case comb::ICmpPredicate::sge:
+              return ">=";
+            }
+            return {};
+          })
+          .Default([](auto) { return ""; });
+  if (!binaryOpcode.empty() && op->getNumOperands() == 2) {
+    auto lhs = emitExpression(op->getOperand(0));
+    auto rhs = emitExpression(op->getOperand(1));
+    if (!lhs || !rhs)
+      return {};
+    return {JObject({
+                {"opcode", binaryOpcode},
+                {"operands", {lhs.expr, rhs.expr}},
+            }),
+            result.getType()};
+  }
+
+  // Expand variadic combinational ops into nested binary HGLDD operations.
+  StringRef variadicOpcode = TypeSwitch<Operation *, StringRef>(op)
+                                 .Case<comb::AndOp>([](auto) { return "&"; })
+                                 .Case<comb::OrOp>([](auto) { return "|"; })
+                                 .Case<comb::XorOp>([](auto) { return "^"; })
+                                 .Default([](auto) { return ""; });
```

Comment:
I think it's inconsistent that `AddOp` and `AndOp` are handled separately. All variadic combinatorial ops (except that concat op) should have been lowered into  binary operations.